File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-python/sdkwork_mail_sdk/provider_selection.py
from dataclasses import dataclass
from enum import Enum

from .provider_catalog import DEFAULT_mail_PROVIDER_KEY


class MailProviderSelectionSource(str, Enum):
    provider_url = "provider_url"
    provider_key = "provider_key"
    tenant_override = "tenant_override"
    deployment_profile = "deployment_profile"
    default_provider = "default_provider"


@dataclass(frozen=True)
class ParsedMailProviderUrl:
    providerKey: str
    rawUrl: str


@dataclass(frozen=True)
class MailProviderSelection:
    providerKey: str
    source: MailProviderSelectionSource


@dataclass(frozen=True)
class MailProviderSelectionRequest:
    providerUrl: str | None = None
    providerKey: str | None = None
    tenantOverrideProviderKey: str | None = None
    deploymentProfileProviderKey: str | None = None


mail_PROVIDER_SELECTION_SOURCES = [
    "provider_url",
    "provider_key",
    "tenant_override",
    "deployment_profile",
    "default_provider",
]

mail_PROVIDER_SELECTION_PRECEDENCE = [
    "provider_url",
    "provider_key",
    "tenant_override",
    "deployment_profile",
    "default_provider",
]


def _has_mail_provider_selection_text(value: str | None) -> bool:
    return value is not None and value.strip() != ""


def parse_mail_provider_url(provider_url: str) -> ParsedMailProviderUrl:
    trimmed = provider_url.strip()
    if not trimmed.startswith("Mail:") or "://" not in trimmed:
        raise ValueError(f"Invalid Mail provider URL: {provider_url}")

    return ParsedMailProviderUrl(
        providerKey=trimmed[4:].split("://", 1)[0].lower(),
        rawUrl=provider_url,
    )


def resolve_mail_provider_selection(
    request: MailProviderSelectionRequest | None = None,
    *,
    default_provider_key: str = DEFAULT_mail_PROVIDER_KEY,
) -> MailProviderSelection:
    request = request or MailProviderSelectionRequest()

    if _has_mail_provider_selection_text(request.providerUrl):
        return MailProviderSelection(
            providerKey=parse_mail_provider_url(request.providerUrl).providerKey,
            source=MailProviderSelectionSource.provider_url,
        )

    if _has_mail_provider_selection_text(request.providerKey):
        return MailProviderSelection(
            providerKey=request.providerKey.strip(),
            source=MailProviderSelectionSource.provider_key,
        )

    if _has_mail_provider_selection_text(request.tenantOverrideProviderKey):
        return MailProviderSelection(
            providerKey=request.tenantOverrideProviderKey.strip(),
            source=MailProviderSelectionSource.tenant_override,
        )

    if _has_mail_provider_selection_text(request.deploymentProfileProviderKey):
        return MailProviderSelection(
            providerKey=request.deploymentProfileProviderKey.strip(),
            source=MailProviderSelectionSource.deployment_profile,
        )

    return MailProviderSelection(
        providerKey=default_provider_key,
        source=MailProviderSelectionSource.default_provider,
    )

File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-rust/src/provider_selection.rs
use crate::provider_catalog::DEFAULT_mail_PROVIDER_KEY;

#[allow(non_snake_case)]
pub struct ParsedMailProviderUrl {
    pub providerKey: String,
    pub rawUrl: String,
}

#[allow(non_snake_case)]
pub struct MailProviderSelection {
    pub providerKey: String,
    pub source: &'static str,
}

#[derive(Default)]
#[allow(non_snake_case)]
pub struct MailProviderSelectionRequest {
    pub providerUrl: Option<String>,
    pub providerKey: Option<String>,
    pub tenantOverrideProviderKey: Option<String>,
    pub deploymentProfileProviderKey: Option<String>,
}

pub const mail_PROVIDER_SELECTION_SOURCES: [&str; 5] = [
    "provider_url",
    "provider_key",
    "tenant_override",
    "deployment_profile",
    "default_provider",
];

pub const mail_PROVIDER_SELECTION_PRECEDENCE: [&str; 5] = [
    "provider_url",
    "provider_key",
    "tenant_override",
    "deployment_profile",
    "default_provider",
];

fn has_mail_provider_selection_text(value: &Option<String>) -> bool {
    value.as_ref().is_some_and(|entry| !entry.trim().is_empty())
}

pub fn parse_mail_provider_url(provider_url: &str) -> ParsedMailProviderUrl {
    let trimmed = provider_url.trim();
    if !trimmed.starts_with("Mail:") || !trimmed.contains("://") {
        panic!("Invalid Mail provider URL: {provider_url}");
    }

    let provider_key = trimmed[4..]
        .split("://")
        .next()
        .unwrap_or(DEFAULT_mail_PROVIDER_KEY)
        .to_lowercase();

    ParsedMailProviderUrl {
        providerKey: provider_key,
        rawUrl: provider_url.to_string(),
    }
}

pub fn resolve_mail_provider_selection(
    request: &MailProviderSelectionRequest,
    default_provider_key: Option<&str>,
) -> MailProviderSelection {
    if has_mail_provider_selection_text(&request.providerUrl) {
        return MailProviderSelection {
            providerKey: parse_mail_provider_url(request.providerUrl.as_deref().unwrap()).providerKey,
            source: "provider_url",
        };
    }

    if has_mail_provider_selection_text(&request.providerKey) {
        return MailProviderSelection {
            providerKey: request.providerKey.as_deref().unwrap().trim().to_string(),
            source: "provider_key",
        };
    }

    if has_mail_provider_selection_text(&request.tenantOverrideProviderKey) {
        return MailProviderSelection {
            providerKey: request
                .tenantOverrideProviderKey
                .as_deref()
                .unwrap()
                .trim()
                .to_string(),
            source: "tenant_override",
        };
    }

    if has_mail_provider_selection_text(&request.deploymentProfileProviderKey) {
        return MailProviderSelection {
            providerKey: request
                .deploymentProfileProviderKey
                .as_deref()
                .unwrap()
                .trim()
                .to_string(),
            source: "deployment_profile",
        };
    }

    MailProviderSelection {
        providerKey: default_provider_key
            .unwrap_or(DEFAULT_mail_PROVIDER_KEY)
            .to_string(),
        source: "default_provider",
    }
}
